en/docs/learn/fraud-detection.md

```diff
@@ -0,0 +1,422 @@
+
+A sample script would look similar to following.
+
+``` javascript
+var onLoginRequest = function(context) {
+    executeStep(1, {
+        onSuccess: function (context) {
+            // Extracting authenticated subject from the first step
+            var user = context.currentKnownSubject;
+            // 'evaluateRiskScore' function will invoke a third party risk based authentication system to analyse the risk score associated with the user.
+            var riskScore = evaluateRiskScore(user);
+            if (riskScore > 7) {
+                Log.info(user.username + ' has a severity high risk score. Hence prompting second factor.');
+                executeStep(2);
+            }
+        }
+    });
+};
+```
+
+#### Custom Post-Authentication Handler
+
+Post authentication handlers are triggered once the authentication is completed, for performing any additional post
+authentication tasks such as authorization. Here also we facilitate a pluggable architecture for custom handlers. You
+ can implement a post authentication handler extending the [AbstractPostAuthnHandler](https://github
+ .com/wso2/carbon-identity-framework/blob/master/components/authentication-framework/org.wso2.carbon.identity
+ .application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework
+ /handler/request/AbstractPostAuthnHandler.java) class. The post-authentication handler interface consists of a
+ single main method which you need to implement in order to implement a post-authentication handler. More details on
+ how to write a custom post-authentication handler found [here](https://is.docs.wso2.com/en/5.10
+ .0/develop/writing-a-post-authentication-handler/).
+
+``` java
+public class CustomPostAuthenticationHandler extends AbstractPostAuthnHandler {
+
+    @Override
+    Public PostAuthnHandlerFlowStatus handle(HttpServletRequest httpServletRequest, HttpServletResponse
+    httpServletResponse, AuthenticationContext authenticationContext) throws PostAuthenticationFailedException {
+
+        AuthenticatedUser user = authenticationContext.getLastAuthenticatedUser();
+        String userName = user.getUserName();
+        Response response = invokeExternalFraudDetectionService(userName);
+        // Evaluate response and decide whether to allow authentication or fail from here.
+    }
+
+    @Override
+    public String getName() {
+        return “CustomPostAuthenticationHandler”;
+    }
+}
+```
+
+#### Custom Authenticator
+
+In Identity Server, an authentication sequence can consist of multiple authentication steps. In this sequence we have
+ the capability to incorporate additional validators in the form of a custom authenticator, in addition to the
+ standard set of authenticators we have. With this way we can secure the authentication flow by ensuring the
+ authenticity of an identity with the help of a third party fraud detection or risk based authentication system.
```

Comment:
In this sequence, it is possible to incorporate additional validators in the form of a custom authenticator in addition to the standard set of authenticators that are available by default. This way, the authentication flow can be secured by ensuring...

Reply:
third-party
risk-based